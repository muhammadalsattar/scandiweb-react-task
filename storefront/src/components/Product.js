import React from "react";
import { connect } from "react-redux";
import Navbar from "./Navbar";
import {addToCart} from "../actions/cart";


export class Product extends React.Component {

  changeFocuesdImage = (e) => {
    const src = e.target.src;
    document.querySelector(".image-focused img").src = src;
  }

  changeActiveAttr = (e) => {
    e.preventDefault();
    document.querySelector(`.${e.target.parentNode.className}#selected`).removeAttribute("id");
    e.target.parentNode.setAttribute("id", "selected");
  }

  addToCart = (product) => {
    let selectedAttr = [];
    document.querySelectorAll(".product-attribute-values div#selected").forEach((node)=>{
      selectedAttr.push({
        name: node.className,
        value: node.children[0].id
      })
    })
    const newProduct = {...product, selectedAttr, quantity: 1}
    this.props.addToCart(newProduct)

    document.querySelector(".cart-overlay").removeAttribute("hidden");
    document.querySelector(".page-body").classList.toggle("faded")
    window.scrollTo({
      top: 0,
      behavior: 'smooth'
    });
  }

  render() {
    return (
      <div className="product-page">
        <Navbar />
        <div className="product page-body">
          <div className="product-gallery">
            <div className="product-gallery-images">
              {this.props.product?.gallery.map((image) => (<img key={image} src={image} onClick={this.changeFocuesdImage} alt={this.props.product.name}/>))}
            </div>
            <div className="image-focused"><img src={this.props.product?.gallery[0]} alt={this.props.product?.name}></img></div>
          </div>
          <div className="product-info">
            <h1 className="product-brand">{this.props.product?.brand}</h1>
            <h1 className="product-name">{this.props.product?.name}</h1>
            <div className="product-attributes">
            {
            this.props.product?.attributes.map(({id, name, items}) =>
              (
              <div className="product-attribute" key={name}>
                <h4>{name}:</h4>
                <div className="product-attribute-values">
                {
                  items.map((item, index) =>
                    name === "Color"?
                      index === 0?
                        <div key={item.id} id="selected" className={id.replaceAll(' ', '-')}><button id={item.value} onClick={this.changeActiveAttr} style={{backgroundColor: item.value}}/></div>:
                        <div key={item.id} className={id.replaceAll(' ', '-')}><button id={item.value} onClick={this.changeActiveAttr} style={{backgroundColor: item.value}}/></div>
                      :
                      index === 0?
                      <div key={item.id} id="selected" className={id.replaceAll(' ', '-')}><button id={item.value} onClick={this.changeActiveAttr}>{item.value}</button></div>: 
                      <div key={item.id} className={id.replaceAll(' ', '-')}><button id={item.value} onClick={this.changeActiveAttr}>{item.value}</button></div>
                  )
                }
                </div>
              </div>
              )
            )}
            </div>
            <div className="product-price">
              <h4>Price:</h4>
              <h4 className="price">
              {this.props.product?.prices.find(({currency: {symbol}}) => symbol === this.props.defaultCurrency.symbol)?.currency.symbol}
              {this.props.product?.prices.find(({currency: {symbol}}) => symbol === this.props.defaultCurrency.symbol)?.amount}
              </h4>
            </div>
            <button className="cart" disabled={!this.props.product?.inStock} onClick={(e)=>{this.addToCart(this.props.product)}}>Add to Cart</button>
            <div className="product-description">
              <p>{new DOMParser().parseFromString(this.props.product?.description, "text/html").body.textContent}</p>
            </div>
          </div>
        </div>
    </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    addToCart: product => dispatch(addToCart(product))
  }
}

const mapStateToProps = ({products, currencies, cart}) => {
  return {
    product: products.products.find(product => product.id === new URLSearchParams(window.location.search).get("id")),
    defaultCurrency: currencies.defaultCurrency,
    cart: cart.cart
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Product);