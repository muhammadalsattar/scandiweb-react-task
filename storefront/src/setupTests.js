const currencies = [
    {
        label: 'USD',
        symbol: '$',
    },
    {
        label: 'GBP',
        symbol: 'Â£',
    },
];
const categories = [
    {
        name: 'category1',
    },
    {
        name: 'category2',
    },
];
const attributes = [
    {
        id: 'Size',
        name: 'Size',
        type: 'text',
        items: [
            {
                id: 40,
                value: '40',
                displayValue: '40',
            },
        ],
    },
    {
        id: 'Color',
        name: 'Color',
        type: 'swatch',
        items: [
            {
                id: 'red',
                value: 'red',
                displayValue: 'Red',
            },
        ],
    },
];
const gallery = [
    'https://dummyimage.com/600x400/000/fff',
    'https://dummyimage.com/600x400/000/ggg',
    'https://dummyimage.com/600x400/000/hhh',
];
const prices = [
    {
        amount: 518.47,
        currency: currencies[0],
    },
    {
        amount: 372.67,
        currency: currencies[1],
    },
];
const products = [
    {
        id: '1',
        name: 'product1',
        brand: 'brand1',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: true,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '40' }],
    },
    {
        id: '2',
        name: 'product2',
        brand: 'brand2',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'red' }],
    },
    {
        id: '3',
        name: 'product3',
        brand: 'brand3',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '41' }],
    },
    {
        id: '4',
        name: 'product4',
        brand: 'brand4',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'orange' }],
    },
    {
        id: '5',
        name: 'product5',
        brand: 'brand5',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: true,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '42' }],
    },
    {
        id: '6',
        name: 'product6',
        brand: 'brand6',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'green' }],
    },
];

export const initialState = {
    currencies,
    categories,
    products,
};
