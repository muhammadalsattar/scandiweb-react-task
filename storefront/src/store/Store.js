import { configureStore, combineReducers } from '@reduxjs/toolkit';
import cartReducer from '../reducers/cartReducer';
import productsReducer from '../reducers/productsReducer';
import categoriesReducer from '../reducers/categoriesReducer';
import currencyReducer from '../reducers/currencyReducer';

const rootReducer = combineReducers({
    cart: cartReducer,
    products: productsReducer,
    categories: categoriesReducer,
    currencies: currencyReducer,
});
// creating store and combining reducers
const Store = configureStore({
    reducer: rootReducer,
});

export default Store;
