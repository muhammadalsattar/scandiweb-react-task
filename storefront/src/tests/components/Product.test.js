import React from "react";
import {render, screen} from "@testing-library/react";
import { Provider } from "react-redux";
import {BrowserRouter as Router} from "react-router-dom";
import Store from "../../store/Store.js";
import {Product} from "../../components/Product.js";
import { initialState } from "../../setupTests.js"

test("Product renders correctly", () => {
    const { undefined, container } = render(
        <Provider store={Store}>
        <Router><Product cart={initialState.cart} defaultCurrency={initialState.currencies[0]} product={initialState.products[0]}/></Router>
        </Provider>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText(/Nike x Stussy/i)).toBeInTheDocument();
    expect(screen.getAllByAltText(/Nike Air Huarache Le/i)).toHaveLength(6);
});