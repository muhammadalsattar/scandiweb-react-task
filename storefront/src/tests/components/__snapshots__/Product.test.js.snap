// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders correctly 1`] = `
<div>
  <div
    class="product-page"
  >
    <nav
      class="navbar"
    >
      <div
        class="categories"
      />
      <div
        class="brand-logo"
      >
        <a
          href="/"
        >
          <img
            alt="logo-icon"
            src="/logo.png"
          />
        </a>
      </div>
      <div
        class="dropdowns"
      >
        <div
          class="currency-div"
        >
          <button
            class="currency-icon"
          >
            <span
              class="currency-drop-icon"
            >
              âˆ§
            </span>
          </button>
          <ul
            class="currency-dropdown"
            hidden=""
          />
        </div>
        <div
          class="cart-div"
        >
          <button
            class="cart-img"
          >
            <img
              alt="cart-img"
              src="/cart.png"
            />
          </button>
          <button
            class="cart-count"
          >
            0
          </button>
          <div
            class="cart-overlay"
            hidden=""
          >
            <p>
              <b>
                My Bag, 
              </b>
              0
               items
            </p>
            <div
              class="cart-total"
            >
              <p>
                Total
              </p>
              <p>
                0.00
              </p>
            </div>
            <div
              class="bag-checkout"
            >
              <div
                class="view-bag"
              >
                <a
                  href="/cart"
                >
                  <button>
                    View Bag
                  </button>
                </a>
              </div>
              <div
                class="checkout"
              >
                <button>
                  Check out
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="product page-body"
    >
      <div
        class="product-gallery"
      >
        <div
          class="product-gallery-images"
        >
          <img
            alt="Nike Air Huarache Le"
            src="https://cdn.shopify.com/s/files/1/0087/6193/3920/products/DD1381200_DEOA_2_720x.jpg?v=1612816087"
          />
          <img
            alt="Nike Air Huarache Le"
            src="https://cdn.shopify.com/s/files/1/0087/6193/3920/products/DD1381200_DEOA_1_720x.jpg?v=1612816087"
          />
          <img
            alt="Nike Air Huarache Le"
            src="https://cdn.shopify.com/s/files/1/0087/6193/3920/products/DD1381200_DEOA_3_720x.jpg?v=1612816087"
          />
          <img
            alt="Nike Air Huarache Le"
            src="https://cdn.shopify.com/s/files/1/0087/6193/3920/products/DD1381200_DEOA_5_720x.jpg?v=1612816087"
          />
          <img
            alt="Nike Air Huarache Le"
            src="https://cdn.shopify.com/s/files/1/0087/6193/3920/products/DD1381200_DEOA_4_720x.jpg?v=1612816087"
          />
        </div>
        <div
          class="image-focused"
        >
          <img
            alt="Nike Air Huarache Le"
            src="https://cdn.shopify.com/s/files/1/0087/6193/3920/products/DD1381200_DEOA_2_720x.jpg?v=1612816087"
          />
        </div>
      </div>
      <div
        class="product-info"
      >
        <h1
          class="product-brand"
        >
          Nike x Stussy
        </h1>
        <h1
          class="product-name"
        >
          Nike Air Huarache Le
        </h1>
        <div
          class="product-attributes"
        >
          <div
            class="product-attribute"
          >
            <h4>
              Size
              :
            </h4>
            <div
              class="product-attribute-values"
            >
              <div
                class="Size"
                id="selected"
              >
                <button
                  id="40"
                >
                  40
                </button>
              </div>
              <div
                class="Size"
              >
                <button
                  id="41"
                >
                  41
                </button>
              </div>
              <div
                class="Size"
              >
                <button
                  id="42"
                >
                  42
                </button>
              </div>
              <div
                class="Size"
              >
                <button
                  id="43"
                >
                  43
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="product-price"
        >
          <h4>
            Price:
          </h4>
          <h4
            class="price"
          >
            $
            144.69
          </h4>
        </div>
        <button
          class="cart"
        >
          Add to Cart
        </button>
        <div
          class="product-description"
        >
          <p>
            Great sneakers for everyday use!
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
